File: zee/src/term_minibuf.c
/* Minibuffer handling
   Copyright (c) 1997-2004 Sandro Sigala.
   Copyright (c) 2003-2005 Reuben Thomas.
   All rights reserved.

   This file is part of Zee.

   Zee is free software; you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2, or (at your option) any later
   version.

   Zee is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with Zee; see the file COPYING.  If not, write to the Free
   Software Foundation, Fifth Floor, 51 Franklin Street, Boston, MA
   02111-1301, USA.  */

#include "config.h"

#include <assert.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "main.h"
#include "extern.h"

void term_minibuf_write(const char *s)
{
  size_t i;

  term_move(term_height() - 1, 0);
  for (i = 0; *s != '\0' && i < term_width(); s++, i++)
    term_addch(*s);
  term_clrtoeol();
}

static void draw_minibuf_read(const char *prompt, const char *value,
                              size_t prompt_len, char *match, size_t pointo)
{
  size_t margin = 1, n = 0;
  size_t width = term_width();

  term_minibuf_write(prompt);

  if (prompt_len + pointo + 1 >= width) {
    margin++;
    term_addch('$');
    n = pointo - pointo % (width - prompt_len - 2);
  }

  term_nprint(min(width - prompt_len - margin, strlen(value) - n), value + n);
  term_print(match);

  if (strlen(value + n) >= width - prompt_len - margin) {
    term_move(term_height() - 1, width - 1);
    term_addch('$');
  }

  term_move(term_height() - 1, prompt_len + margin - 1 +
            pointo % (width - prompt_len - margin));

  term_refresh();
}


/*
 * Minibuffer key action routines
 */

static void mb_suspend(void)
{
  FUNCALL(suspend);
}

static void mb_return(void)
{
  term_move(term_height() - 1, 0);
  term_clrtoeol();
}

static void mb_cancel(void)
{
  term_move(term_height() - 1, 0);
  term_clrtoeol();
}

static ptrdiff_t mb_bol(void)
{
  ptrdiff_t i;
  i = 0;
  return(i);
}

static ptrdiff_t mb_eol(astr as)
{
  ptrdiff_t i;
  i = astr_len(as);
  return(i);
}

static ptrdiff_t mb_backward_char(ptrdiff_t i)
{
  if (i > 0)
    --i;
  else
    ding();
  return(i);
}

static ptrdiff_t mb_forward_char(ptrdiff_t i, astr as)
{
  if ((size_t)i < astr_len(as))
    ++i;
  else
    ding();
  return(i);
}

static void mb_kill_line(ptrdiff_t i, astr as)
{
  if ((size_t)i < astr_len(as))
    astr_truncate(as, i);
  else
    ding();
}

static ptrdiff_t mb_backward_delete_char(ptrdiff_t i, astr as)
{
  if (i > 0)
    astr_remove(as, --i, 1);
  else
    ding();
  return(i);
}

static void mb_delete_char(ptrdiff_t i, astr as)
{
  if ((size_t)i < astr_len(as))
    astr_remove(as, i, 1);
  else
    ding();
}

static int mb_scroll_down(Completion *cp, int thistab, int lasttab)
{
  if (cp == NULL)
    ding();
  else if (cp->flags & COMPLETION_POPPEDUP) {
    popup_scroll_down();
    thistab = lasttab;
  }
  return(thistab);
}

static int mb_scroll_up(Completion *cp, int thistab, int lasttab)
{
  if (cp == NULL)
    ding();
  else if (cp->flags & COMPLETION_POPPEDUP) {
    popup_scroll_up();
    thistab = lasttab;
  }
  return(thistab);
}

static void mb_prev_history(History *hp, astr as, ptrdiff_t *_i, char **_saved)
{
  ptrdiff_t i = *_i;
  char *saved = *_saved;
  if (hp) {
    const char *elem = previous_history_element(hp);
    if (elem) {
      if (!saved)
        saved = zstrdup(astr_cstr(as));

      i = strlen(elem);
      astr_cpy_cstr(as, elem);
    }
  }
  *_i = i;
  *_saved = saved;
}

static void mb_next_history(History *hp, astr as, ptrdiff_t *_i, char **_saved)
{
  ptrdiff_t i = *_i;
  char *saved = *_saved;
  if (hp) {
    const char *elem = next_history_element(hp);
    if (elem) {
      i = strlen(elem);
      astr_cpy_cstr(as, elem);
    }
    else if (saved) {
      i = strlen(saved);
      astr_cpy_cstr(as, saved);
      saved = NULL;
    }
  }
  *_i = i;
  *_saved = saved;
}

static void mb_complete(Completion *cp, int lasttab, astr as, int *_thistab, ptrdiff_t *_i)
{
  int thistab = *_thistab;
  ptrdiff_t i = *_i;
  if (cp == NULL)
    ding();
  else {
    if (lasttab != COMPLETION_NOTCOMPLETING &&
        lasttab != COMPLETION_NOTMATCHED &&
        cp->flags & COMPLETION_POPPEDUP) {
      popup_scroll_up();
      thistab = lasttab;
    } else {
      astr bs = astr_new();
      astr_cpy(bs, as);
      thistab = completion_try(cp, bs, TRUE);
      assert(thistab != COMPLETION_NOTCOMPLETING);
      switch (thistab) {
      case COMPLETION_NONUNIQUE:
      case COMPLETION_MATCHED:
      case COMPLETION_MATCHEDNONUNIQUE:
        i = cp->matchsize;
        bs = astr_new();
        astr_ncat(bs, cp->match, cp->matchsize);
        if (astr_cmp(as, bs) != 0)
          thistab = COMPLETION_NOTCOMPLETING;
        astr_cpy(as, bs);
        break;
      case COMPLETION_NOTMATCHED:
        ding();
      }
    }
  }
  *_thistab = thistab;
  *_i = i;
}

static ptrdiff_t mb_self_insert(int c, ptrdiff_t i, astr as)
{
  if (c > 255 || !isprint(c))
    ding();
  else
    astr_insert_char(as, i++, c);
  return i;
}

/*
 * Read a string from the minibuffer.
 */
astr term_minibuf_read(const char *prompt, const char *value, Completion *cp, History *hp)
{
  int c, thistab, lasttab = COMPLETION_NOTCOMPLETING, ret = FALSE;
  ptrdiff_t i;
  char *s[] = {"", " [No match]", " [Sole completion]", " [Complete, but not unique]", ""};
  char *saved = NULL;
  astr as = astr_new(), retval = NULL;

  if (hp)
    prepare_history(hp);

  i = strlen(value);
  astr_cpy_cstr(as, value);

  for (;;) {
    draw_minibuf_read(prompt, astr_cstr(as), strlen(prompt), s[lasttab], (size_t)i);

    thistab = COMPLETION_NOTCOMPLETING;

    switch (c = getkey()) {
    case KBD_NOKEY:
      break;
    case KBD_CTRL | 'z':
      mb_suspend();
      break;
    case KBD_RET:
      mb_return();
      retval = as;
      ret = TRUE;
      break;
    case KBD_CANCEL:
      mb_cancel();
      ret = TRUE;
      break;
    case KBD_CTRL | 'a':
    case KBD_HOME:
      i = mb_bol();
      break;
    case KBD_CTRL | 'e':
    case KBD_END:
      i = mb_eol(as);
      break;
    case KBD_CTRL | 'b':
    case KBD_LEFT:
      i = mb_backward_char(i);
      break;
    case KBD_CTRL | 'f':
    case KBD_RIGHT:
      i = mb_forward_char(i, as);
      break;
    case KBD_CTRL | 'k':
      mb_kill_line(i, as);
      break;
    case KBD_BS:
      i = mb_backward_delete_char(i, as);
      break;
    case KBD_DEL:
      mb_delete_char(i, as);
      break;
    case KBD_META | 'v':
    case KBD_PGUP:
      thistab = mb_scroll_down(cp, thistab, lasttab);
      break;
    case KBD_CTRL | 'v':
    case KBD_PGDN:
      thistab = mb_scroll_up(cp, thistab, lasttab);
      break;
    case KBD_UP:
    case KBD_META | 'p':
      mb_prev_history(hp, as, &i, &saved);
      break;
    case KBD_DOWN:
    case KBD_META | 'n':
      mb_next_history(hp, as, &i, &saved);
      break;
    case KBD_TAB:
      mb_complete(cp, lasttab, as, &thistab, &i);
      break;
    default:
      i = mb_self_insert(c, i, as);
    }

    lasttab = thistab;
    if (ret)
      break;
  }

  if (cp && cp->flags & COMPLETION_POPPEDUP) {
    popup_clear();
    term_refresh();
  }

  return retval;
}

File: src/term_minibuf.lua
-- Minibuffer handling
--
-- Copyright (c) 2010-2012 Free Software Foundation, Inc.
--
-- This file is part of Zee.
--
-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 3, or (at your option)
-- any later version.
--
-- This program is distributed in the hope that it will be useful, but
-- WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

local function draw_minibuf_read (prompt, value, pointo)
  term_minibuf_write (prompt)

  local w, h = term_width (), term_height ()
  local margin = 1
  local n = 0

  if #prompt + pointo + 1 >= w then
    margin = margin + 1
    term_addstr ('$')
    n = pointo - pointo % (w - #prompt - 2)
  end

  term_addstr (string.sub (value, n + 1, math.min (w - #prompt - margin, #value - n)))

  if #value - n >= w - #prompt - margin then
    term_move (h - 1, w - 1)
    term_addstr ('$')
  end

  term_move (h - 1, #prompt + margin - 1 + pointo % (w - #prompt - margin))

  term_refresh ()
end

function term_minibuf_read (prompt, value, pos, cp, hp)
  local quit = false

  Define ("minibuf-insert-character", "",
          function ()
            local key = term_keytobyte (lastkey ())
            if not key then
              ding ()
              return false
            end
            value = value:sub (1, pos) .. string.char (key) .. value:sub (pos + 1)
            pos = pos + 1
          end)

  Define ("minibuf-exit", "",
          function ()
            quit = true
          end)

  Define ("minibuf-suspend", "",
          function ()
            execute_command ("file-suspend")
          end)

  Define ("minibuf-quit", "",
          function ()
            value = nil
            quit = true
          end)

  Define ("minibuf-move-start-line", "",
          function ()
            pos = 0
          end)

  Define ("minibuf-move-end-line", "",
          function ()
            pos = #value
          end)

  Define ("minibuf-move-previous-character", "",
          function ()
            if pos > 0 then
              pos = pos - 1
            else
              ding ()
            end
          end)

  Define ("minibuf-move-next-character", "",
          function ()
            if pos < #value then
              pos = pos + 1
            else
              ding ()
            end
          end)

  Define ("minibuf-delete-previous-character", "",
          function ()
            if pos > 0 then
              value = value:sub (1, pos - 1) .. value:sub (pos + 1)
              pos = pos - 1
            else
              ding ()
            end
          end)

  Define ("minibuf-delete-next-character", "",
          function ()
            if pos < #value then
              value = value:sub (1, pos) .. value:sub (pos + 2)
            else
              ding ()
            end
          end)

  Define ("minibuf-previous-page", "",
          function ()
            if cp == nil then
              ding ()
            elseif cp.poppedup then
              popup_scroll_down ()
            end
          end)

  Define ("minibuf-next-page", "",
          function ()
            if cp == nil then
              ding ()
            elseif cp.poppedup then
              popup_scroll_up ()
            end
          end)

  Define ("minibuf-history-previous", "",
          function ()
            if hp then
              local elem = previous_history_element (hp)
              if elem then
                if not saved then
                  saved = value
                end
                value = elem
              end
            end
          end)

  Define ("minibuf-history-next", "",
          function ()
            if hp then
              local elem = next_history_element (hp)
              if elem then
                value = elem
              elseif saved then
                value = saved
                saved = nil
              end
            end
          end)

  Define ("minibuf-complete", "",
          function ()
            if not cp or #cp.matches == 0 then
              ding ()
            else
              if value ~= cp.match then
                value = cp.match
                pos = #value
              else
                popup_scroll_down_and_loop ()
              end
            end
          end)

  local minibuf_bindings = {}
  bindings = minibuf_bindings
  bind_printing_chars ("minibuf-insert-character")
  key_bind ("RET", "minibuf-exit")
  key_bind ("C-z", "minibuf-suspend")
  key_bind ("C-g", "minibuf-quit")
  key_bind ("C-a", "minibuf-move-start-line")
  key_bind ("HOME", "minibuf-move-start-line")
  key_bind ("C-e", "minibuf-move-end-line")
  key_bind ("END", "minibuf-move-end-line")
  key_bind ("C-b", "minibuf-move-previous-character")
  key_bind ("LEFT", "minibuf-move-previous-character")
  key_bind ("C-f", "minibuf-move-next-character")
  key_bind ("RIGHT", "minibuf-move-next-character")
  key_bind ("BACKSPACE", "minibuf-delete-previous-character")
  key_bind ("C-d", "minibuf-delete-next-character")
  key_bind ("DELETE", "minibuf-delete-next-character")
  key_bind ("M-v", "minibuf-previous-page")
  key_bind ("PAGEUP", "minibuf-previous-page")
  key_bind ("C-v", "minibuf-next-page")
  key_bind ("PAGEDOWN", "minibuf-next-page")
  key_bind ("M-p", "minibuf-history-previous")
  key_bind ("UP", "minibuf-history-previous")
  key_bind ("M-n", "minibuf-history-next")
  key_bind ("DOWN", "minibuf-history-next")
  key_bind ("TAB", "minibuf-complete")

  if hp then
    history_prepare (hp)
  end

  local saved
  pos = pos or #value

  local completion_text, old_completion_text
  repeat
    if cp and (not old_completion_text or old_completion_text ~= completion_text) then
      popup_completion (cp)
      completion_try (cp, value)
      local completion_text = completion_write (cp, win.ewidth)
      popup_set (completion_text)
      old_completion_text = completion_text
    end

    draw_minibuf_read (prompt, value, pos)
    local key = get_key_chord (true)
    local name = get_command_by_key (key)
    if command_exists (name) then
      call_command (name)
    else
      ding ()
    end
  until quit

  minibuf_clear ()
  if cp then
    popup_clear ()
    term_redisplay ()
  end
  bindings = root_bindings
  return value
end

function term_minibuf_write (s)
  term_move (term_height () - 1, 0)
  term_clrtoeol ()
  term_addstr (string.sub (s, 1, math.min (#s, term_width ())))
end
